File: tests/assets/framework/src/codegen/plant/plant.c
#include "src/model/plant.h"
#include <pthread>
struct Plant_t init() {
    struct Plant_t plant;
    plant.base = (char *) malloc(sizeof(char) * 16);
    plant.systems = (struct System_t *) malloc(sizeof(struct System_t) * <palinka.utils.Database 
    object 
        at0x000002D3F2329788>); 
    plant.nb_systems = <palinka.utils.Database object at 0x000002D3F2329788>;
    plant.systems[0]->offset = 0;
    plant.systems[0]->size = 2;
    pthread_rwlock_init(& plant.systems[0]->rwlock);
    plant.systems[0]->step = sys_net01_step;
    sys_net01_init(plant.systems[0]);
    plant.systems[1]->offset = 2;
    plant.systems[1]->size = 1;
    pthread_rwlock_init(& plant.systems[1]->rwlock);
    plant.systems[1]->step = sys_sys01_io_01_step;
    sys_sys01_io_01_init(plant.systems[1]);
    plant.systems[2]->offset = 3;
    plant.systems[2]->size = 1;
    pthread_rwlock_init(& plant.systems[2]->rwlock);
    plant.systems[2]->step = sys_sys01_im_step;
    sys_sys01_im_init(plant.systems[2]);
    plant.systems[3]->offset = 4;
    plant.systems[3]->size = 4;
    pthread_rwlock_init(& plant.systems[3]->rwlock);
    plant.systems[3]->step = sys_sys01_step;
    sys_sys01_init(plant.systems[3]);
    plant.systems[4]->offset = 8;
    plant.systems[4]->size = 6;
    pthread_rwlock_init(& plant.systems[4]->rwlock);
    plant.systems[4]->step = sys_sys02_step;
    sys_sys02_init(plant.systems[4]);
    plant.systems[5]->offset = 14;
    plant.systems[5]->size = 2;
    pthread_rwlock_init(& plant.systems[5]->rwlock);
    plant.systems[5]->step = sys_sys03_step;
    sys_sys03_init(plant.systems[5]);
    return plant;
}

File: palinka/ast/automation/plant.py
from ...model.automation import Plant
from ...model import ast
from . import system

import palinka.ast.utils as astu


def build(plant: Plant):
    systems: list[ast.automation.SystemDeclaration] = [system.build(sys) for sys in plant.get_systems()]
    
    preprocessors =  [
        ast.Preprocessor(
            ast.Include(ast.Identifier("src/model/plant.h"), internal=True)
        ),
        ast.Preprocessor(
            ast.Include(ast.Identifier("pthread"))
        )
    ]
    
    return ast.automation.PlantDeclaration(
        systems, 
        ast.automation.TranslationUnit([
            build_init_function(plant)
        ],preprocessors)
    )

def build_init_function(plant: Plant):
    declarations = []
    statements = []

    declarations.append(
        astu.struct_var_decl("Plant_t", "plant", as_ptr=False)
    )

    statements += [astu.assign_stmt(
        astu.attr_access_expr("plant", "base", arrow=False),
        astu.cast_expr(
            astu.typename("char", as_ptr=True),     
            astu.function_call_expr(
                "malloc",
                astu.mult_expr(
                    astu.sizeof(
                        ast.TypeName.create(
                            [ast.Char().as_type_specifier().as_declaration_specifier()],
                            None
                        )
                    ),
                    astu.id_expr("@plant")
                )
            )
        )
    )]

    nb_systems = len(list(iter(plant.get_systems())))

    statements += [astu.assign_stmt(
        astu.attr_access_expr("plant", "systems", arrow=False),
        astu.cast_expr(
            astu.typename(
                astu.struct_specifier("System_t").as_type_specifier(),
                as_ptr=True
            ),   
            astu.function_call_expr(
                "malloc",
                astu.mult_expr(
                    astu.sizeof(
                        ast.TypeName.create(
                            [astu.struct_specifier("System_t").as_type_specifier().as_declaration_specifier()],
                            None
                        )
                    ),
                    astu.id_expr("&plant")
                )
                
            )
        )
    )]

    statements += [astu.assign_stmt(
         astu.attr_access_expr("plant", "nb_systems", arrow=False),
         astu.id_expr("&plant")
    )]

    for i, system in enumerate(plant.get_systems()):
        sys_namespace = f"sys_{system.get_slug_id()}"

        statements += [astu.assign_stmt(
            astu.attr_access_expr(
                astu.getitem_expr(
                    astu.attr_access_expr("plant", "systems", arrow=False),
                    astu.id_expr(f"#{system.get_id()}")
                ),
                "offset"
            ),
            astu.id_expr(f"${system.get_id()}")
        ), astu.assign_stmt(
            astu.attr_access_expr(
                astu.getitem_expr(
                    astu.attr_access_expr("plant", "systems", arrow=False),
                    astu.id_expr(f"#{system.get_id()}")
                ),
                "size"
            ),
            astu.id_expr(f"@{system.get_id()}")
        ), astu.function_call_stmt(
            "pthread_rwlock_init",
            astu.ref_expr(
                astu.attr_access_expr(
                    astu.getitem_expr(
                        astu.attr_access_expr("plant", "systems", arrow=False),
                        astu.id_expr(f"#{system.get_id()}")
                    ),
                    "rwlock"
                )
            )

        ), astu.assign_stmt(
            astu.attr_access_expr(
                astu.getitem_expr(
                    astu.attr_access_expr("plant", "systems", arrow=False),
                    astu.id_expr(f"#{system.get_id()}")
                ),
                "step"
            ),
            astu.id_expr(f"{sys_namespace}_step")
        ), astu.function_call_stmt(
            f"{sys_namespace}_init",
            astu.getitem_expr(
                    astu.attr_access_expr("plant", "systems", arrow=False),
                    astu.id_expr(f"#{system.get_id()}")
            )
        )]        

    statements += [astu.return_stmt("plant")]
    
    return astu.func_decl(
        [astu.struct_specifier("Plant_t")],
        astu.func_declarator(
            "init",
            None,
            as_ptr=False
        ),
        [],
        ast.CompoundStatement.create(declarations, statements)
    )